### PR TITLE
slack connector: implemented signing secret and updated docs

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- validate slack requests using signing secret
- updated documentation page (old instructions were outdated, added screenshots)
- moved slack tests into their own module (added new tests for signature at the bottom)

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: docs/docs/connectors/slack.mdx [hunk 1]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
+
+   <Image img={eventsImg} caption="Subscribe to Bot Events" alt="Subscribe to Bot Events Screenshot" max-width="500px"/>
+
+   Make sure to hit **Save Changes** at the bottom of the page after you've added these events.
+
+   (If you missed to grant all needed permissions to your app while setting up
+   sending of messages, you'll be prompted to **reinstall your app** which you will
+   need to do. Otherwise, Slack will confirm your change with a **Success!**)
+
+Your bot is now ready to go and will receive webhook notifications about new mewssages.
```
Comment (tmbo) [inline docs/docs/connectors/slack.mdx:140]:
```suggestion
Your bot is now ready to go and will receive webhook notifications about new messages.
```

**Comment by tmbo** [comment]
@erohmensing added images to the slack docs, as it was very hard to follow them without. does this look sane?

File: docs/docs/connectors/slack.mdx [hunk 2]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
-You first have to create a Slack app to get credentials.
-Once you have them you can add these to your `credentials.yml`.
+Connecting a bot to Slack requires you to configure it to send messages
+(using API credentials) and to receive messages (using a webhook).
 
-## Getting Credentials
+## Sending Messages
 
-**How to get the Slack credentials:** You need to set up a Slack app.
+Create a new file `credentials.yml` in the root folder of your rasa open source
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx:20]:
```suggestion
Create a new file `credentials.yml` in the root folder of your Rasa Open Source
```

File: docs/docs/connectors/slack.mdx [hunk 3]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
-* Interactivity & Shortcuts > Interactivity (if your bot uses any
-  [interactive components](https://api.slack.com/reference/block-kit/interactive-components "The Interactive Components section of your Slack interface") , e.g. buttons)
+4. Head over to **Basic Information** to gather the **Signing Secret**.
 
-* Add your Rasa request URL `http://<host>:<port>/webhooks/slack/webhook`, replacing
-  the host and port with the appropriate values from your running Rasa X or Rasa Open Source server.
+   <Image img={secretImg} caption="Signing Secret" alt="Signing Secret Screenshot" max-width="500px"/>
 
-* Event subscriptions > Subscribe to bot events: `message.channels`, `message.groups`, `message.im`, `message.mpim`
+   Copy the signing secret into your `configuration.yml` as the value for `slack_signing_secret`:
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx:86]:
```suggestion
   Copy the signing secret into your `credentials.yml` as the value for `slack_signing_secret`:
```

File: docs/docs/connectors/slack.mdx [hunk 4]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
+Make sure you've gone through [Sending Messages](./slack.mdx#sending-messages) first
+and you have a ready to go `credentials.yml`.
 
-3. Configure your `credentials.yml` with the new Slack credentials as described in the
-   [Running on Slack](./slack.mdx#running-on-slack) section, and restart your server. This will allow
-   your server to confirm that the endpoint is ready to receive events. Once you've restarted the server,
-   go to the “Event Subscriptions” section, turn on the “Enable Events” and add the endpoint as the Request URL.
+To receive messages, you will need a publicly available URL for Slack to reach
+your bot and tell you about new messages. If you are running locally, you can
+[test channels using ngrok](../messaging-and-voice-channels#testing-channels-on-your-local-machine)
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx]:
```suggestion
[test channels using ngrok](../messaging-and-voice-channels#testing-channels-on-your-local-machine).
```

File: docs/docs/connectors/slack.mdx [hunk 5]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
-   to provide this value in your credentials later in these instructions. It should start
-   with `xoxb`.
+This setup will allow your bot to send messages. Now let's head over to the setup
+for receiving and reacting to messages.
 
-2. In the “OAuth & Permissions > Redirect URLs” enter the endpoint for receiving Slack messages. This is
-   the same URL you entered above for Interactivity & Shortcuts - `http://<host>:<port>/webhooks/slack/webhook`.
+## Receiving Messages
+Make sure you've gone through [Sending Messages](./slack.mdx#sending-messages) first
+and you have a ready to go `credentials.yml`.
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx:100]:
```suggestion
Before continuing, make sure you have configured a Slack App for [Sending Messages](./slack.mdx#sending-messages) and have added Slack credentials to your `credentials.yml` file.
```

File: docs/docs/connectors/slack.mdx [hunk 6]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
+   - `message.channels`,
+   - `message.groups`,
+   - `message.im` and
+   - `message.mpim`.
+
+   <Image img={eventsImg} caption="Subscribe to Bot Events" alt="Subscribe to Bot Events Screenshot" max-width="500px"/>
+
+   Make sure to hit **Save Changes** at the bottom of the page after you've added these events.
+
+   (If you missed to grant all needed permissions to your app while setting up
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx:136]:
```suggestion
   (If you didn't grant all required permissions to your app while setting up
```

File: docs/docs/connectors/slack.mdx [hunk 7]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
+   sending of messages, you'll be prompted to **reinstall your app** which you will
+   need to do. Otherwise, Slack will confirm your change with a **Success!**)
+
+Your bot is now ready to go and will receive webhook notifications about new messages.
+## Optional: Interactive Components
+After you've completed [Sending Messages](./slack.mdx#sending-messages) and
+[Receiving Messages](./slack.mdx#receiving-messages) your bot is ready to go. If you
+want to use Slacks "Interactive" elements (buttons or menus), you'll need to do
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx:144]:
```suggestion
want to use Slack's [interactive components](https://api.slack.com/block-kit/interactivity) (buttons or menus), you'll need to do
```

File: docs/docs/connectors/slack.mdx [hunk 8]
```diff
@@ -4,69 +4,168 @@ sidebar_label: Slack
...
+some additional configuration:
+
+Open the **Interactivity & Shortcuts** page and toggle **Interactivity** to be on.
+Afterwards, you'll need to enter the same url into the **Request URL** field which
+you used in Step 2. of Receiving Messages, e.g. `https://<host>/webhooks/slack/webhook`.
```
Comment (erohmensing) [inline docs/docs/connectors/slack.mdx:149]:
```suggestion
you used in Step 2 of [Receiving Messages](./slack.mdx#receiving-messages), e.g. `https://<host>/webhooks/slack/webhook`.
```

**Review by erohmensing — COMMENTED** [review]
I think the screenshots are very helpful!

File: rasa/core/channels/slack.py [hunk 9]
```diff
@@ -399,7 +453,16 @@ async def health(_: Request) -> HTTPResponse:
         @slack_webhook.route("/webhook", methods=["GET", "POST"])
         async def webhook(request: Request) -> HTTPResponse:
             content_type = request.headers.get("content-type")
-            # Slack API sends either a JSON-encoded or a URL-encoded body depending on the content
+
+            if not self.is_request_from_slack_authentic(request):
```
Comment (joejuzl) [inline rasa/core/channels/slack.py:457]: Is this the only channel that does message request level authentication/validation? And if not, is it always handled directly in the blueprint?

File: rasa/core/channels/slack.py [hunk 10]
```diff
@@ -387,6 +412,35 @@ def get_metadata(self, request: Request) -> Dict[Text, Any]:
 
         return {}
 
+    def is_request_from_slack_authentic(self, request: Request) -> bool:
```
Comment (joejuzl) [inline rasa/core/channels/slack.py:415]: Could there be a unit test that can assert this correctly blocks a request with an incorrect signature?

File: rasa/core/channels/slack.py [hunk 11]
```diff
@@ -183,6 +194,20 @@ def __init__(
...
         self.retry_num_header = slack_retry_number_header
         self.use_threads = use_threads
+        self.slack_signing_secret = slack_signing_secret
+
+        self._raise_deprecation_warnings()
+
+    def _raise_deprecation_warnings(self) -> None:
+        """Raises any deprecation warning regarding configuration parameters."""
+        if self.slack_signing_secret is None:
+            rasa.shared.utils.io.raise_warning(
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/channels/slack.py:204]: good point 😂 :+1:

[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/channels/slack.py:204]: wait, no there is a reason: we do have two different paths of deprecation, `FutureWarning` and `DeprecationWarning`. Ther is a slight difference in how they are handled and their semantics, basically A `FutureWarning` is aimed at an "end user", so someone running our CLI entry points, e.g. `rasa train`. If we deprecate a configuration setting, or in this case add the requirement for a configuration setting we need to alert this end user to this change and hence need to log a warning (only done for `FutureWarning` by default). `DeprecationWarning` is aimed at other developers customizing rasa open source (e.g. writing their own NLU components), in this case we expect them to have written custom library code that uses our library and in this case we use the default deprecation practices in python.

[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/channels/slack.py:204]: https://www.notion.so/rasa/Python-Code-Conventions-26835ffd72484dd699f28fa693916d2e#6d4bea523bef4b18b200f47988087d4e

[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/channels/slack.py:204]: feel free to ignore the above, all right but I got confused, `raise_deprecation_warning` is actually the right one and raises a `FutureWarning`

[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/core/channels/slack.py:204]: which leads me to think that we should have chosen better names for these functions...

File: rasa/core/channels/slack.py (hunk 9 above)
Comment (tmbo) [reply rasa/core/channels/slack.py:457]: I don't think all different channels have this functionality and some choose a different way (e.g. telegram recommends to choose an "unguessable" URL, e.g. https://www.example.com/<token>). We'll need to secure a couple more channels, it seems to be implemented for facebook, but I didn't yet check all of them

Comment (joejuzl) [reply rasa/core/channels/slack.py:458]: ok, makes sense. Maybe when a few more are secured we could consider making a standard way of doing it e.g. an overridable abstract method on `InputChannel` that is called every time.

File: rasa/core/channels/slack.py [hunk 12]
```diff
@@ -166,14 +171,20 @@ def __init__(
                 If not set, messages will be sent back
                 to the "App" DM channel of your bot's name.
             proxy: A Proxy Server to route your traffic through
-            slack_retry_reason_header: Slack HTTP header name indicating reason that slack send retry request.
-            slack_retry_number_header: Slack HTTP header name indicating the attempt number
+            slack_retry_reason_header: Slack HTTP header name indicating reason
+                that slack send retry request.
+            slack_retry_number_header: Slack HTTP header name indicating
+                the attempt number
```
Comment (wochinge) [inline rasa/core/channels/slack.py:177]:
```suggestion
                the attempt number.
```

File: rasa/core/channels/slack.py [hunk 13]
```diff
@@ -166,14 +171,20 @@ def __init__(
...
             errors_ignore_retry: Any error codes given by Slack
                 included in this list will be ignored.
                 Error codes are listed
                 `here <https://api.slack.com/events-api#errors>`_.
-            use_threads: If set to True, your bot will send responses in Slack as a threaded message.
-                Responses will appear as a normal Slack message if set to False.
+            use_threads: If set to True, your bot will send responses in Slack as
```
Comment (wochinge) [inline rasa/core/channels/slack.py:182]:
```suggestion
            use_threads: If set to `True`, your bot will send responses in Slack as
```

File: rasa/core/channels/slack.py [hunk 14]
```diff
@@ -166,14 +171,20 @@ def __init__(
...
+                a threaded message. Responses will appear as a normal Slack message
+                if set to False.
```
Comment (wochinge) [inline rasa/core/channels/slack.py:184]:
```suggestion
                if set to `False`.
```

File: rasa/core/channels/slack.py [hunk 15]
```diff
@@ -166,14 +171,20 @@ def __init__(
...
+            slack_signing_secret: Slack creates a unique string for your app and
+                shares it with you. This allows us to verify requests from Slack
+                with confidence by verifying signatures using your signing secret.
```
Comment (wochinge) [inline rasa/core/channels/slack.py:187]: optional: what does the `with confidence` part mean?

File: rasa/core/channels/slack.py [hunk 16]
```diff
@@ -211,7 +237,7 @@ def _is_user_message(slack_event: Dict) -> bool:
         )
 
     @staticmethod
-    def _sanitize_user_message(text, uids_to_remove) -> Text:
+    def _sanitize_user_message(text, uids_to_remove: Optional[List[Text]]) -> Text:
```
Comment (wochinge) [inline rasa/core/channels/slack.py:240]: thanks for all the typing!
```suggestion
    def _sanitize_user_message(text: Text, uids_to_remove: Optional[List[Text]]) -> Text:
```

File: rasa/core/channels/slack.py [hunk 17]
```diff
@@ -387,6 +416,35 @@ def get_metadata(self, request: Request) -> Dict[Text, Any]:
...
+        we don't process request from a third-party disguising as slack.
+
+        Args:
+            request: incoming request to be checked
+
+        Returns:
+            `True` if the request came from Slack.
+        """
+        if self.slack_signing_secret is None:
+            # this is done for backwards compatibility, but will be removed in
```
Comment (wochinge) [inline rasa/core/channels/slack.py:432]: optional: Should we add a todo we don't forget? And do you create the removal PR against `master`? Alternatively add an issue to the milestone please.
```suggestion
            # TODO: Remove when releasing 2.1.0
            # this is done for backwards compatibility, but will be removed in
```

File: rasa/core/channels/slack.py [hunk 18]
```diff
@@ -387,6 +416,35 @@ def get_metadata(self, request: Request) -> Dict[Text, Any]:
...
+            # the next release
+            return True
+
+        slack_signing_secret = bytes(self.slack_signing_secret, "utf-8")
+
+        slack_signature = request.headers.get("X-Slack-Signature", "")
+        slack_request_timestamp = request.headers.get("X-Slack-Request-Timestamp", "")
```
Comment (wochinge) [inline rasa/core/channels/slack.py:439]: Should we also do step 2 of this guide https://api.slack.com/authentication/verifying-requests-from-slack#verifying-requests-from-slack-using-signing-secrets__a-recipe-for-security__step-by-step-walk-through-for-validating-a-request (avoid replay attacks)?

File: rasa/core/channels/slack.py [hunk 19]
```diff
@@ -399,7 +457,16 @@ async def health(_: Request) -> HTTPResponse:
         @slack_webhook.route("/webhook", methods=["GET", "POST"])
         async def webhook(request: Request) -> HTTPResponse:
             content_type = request.headers.get("content-type")
-            # Slack API sends either a JSON-encoded or a URL-encoded body depending on the content
+
+            if not self.is_request_from_slack_authentic(request):
+                return response.text(
+                    "Message is not properly signed with a valid "
+                    "X-Slack-Signature header",
```
Comment (wochinge) [inline rasa/core/channels/slack.py:464]: should we include a link to the docs?

File: rasa/core/channels/slack.py [hunk 20]
```diff
@@ -399,7 +457,16 @@ async def health(_: Request) -> HTTPResponse:
...
+                    status=400,
```
Comment (wochinge) [inline rasa/core/channels/slack.py:465]:
```suggestion
                    status=HTTPStatus.BAD_REQUEST,
```

Comment (wochinge) [inline rasa/core/channels/slack.py:465]: this seems to fit better, what do you think?
```suggestion
                    status=HTTPStatus.UNAUTHORIZED,
```

File: rasa/core/channels/slack.py [hunk 21]
```diff
@@ -412,23 +479,22 @@ async def webhook(request: Request) -> HTTPResponse:
...
-                        on_new_message,
-                        text=self._sanitize_user_message(
-                            user_message, metadata["users"]
-                        ),
-                        sender_id=sender_id,
-                        metadata=metadata,
-                    )
-                else:
+                if not self._is_user_message(output):
+                    return response.text("Bot message delivered.")
```
Comment (wochinge) [inline rasa/core/channels/slack.py:483]: Should we add a `logger.debug` message for this case?

File: tests/core/channels/test_slack.py [hunk 22]
```diff
@@ -0,0 +1,798 @@
...
+
+    bot = SlackBot("DummyToken", "General", thread_id="DummyThread")
+    assert bot.client.token == "DummyToken"
+    assert bot.slack_channel == "General"
+    assert bot.thread_id == "DummyThread"
+
+
+# Use monkeypatch for sending attachments, images and plain text.
+@pytest.mark.filterwarnings("ignore:unclosed.*:ResourceWarning")
+@pytest.mark.asyncio
```
Comment (wochinge) [inline tests/core/channels/test_slack.py:370]: do we actually need `@pytest.mark.asyncio`?

File: tests/core/channels/test_slack.py [hunk 23]
```diff
@@ -0,0 +1,798 @@
...
+
+        assert request_params == {
+            "as_user": True,
+            "channel": "General",
+            "thread_ts": "DummyThread",
+            "test_key": "test_value",
+        }
+
+
+def prepare_slack_request(headers):
```
Comment (wochinge) [inline tests/core/channels/test_slack.py:749]: types please

File: tests/core/channels/test_slack.py [hunk 24]
```diff
@@ -0,0 +1,798 @@
...
+            "185b90342d467abf5c74a27d2d0dd1f5",
+            "x-slack-request-timestamp": "1604586653",
+        }
+    )
+    input_with_right_secret = SlackInput("mytoken", slack_signing_secret="foobar")
+    input_with_wrong_secret = SlackInput("mytoken", slack_signing_secret="foobaz")
+
+    assert input_with_right_secret.is_request_from_slack_authentic(request) is True
+
+    assert input_with_wrong_secret.is_request_from_slack_authentic(request) is False
```
Comment (wochinge) [inline tests/core/channels/test_slack.py:787]: optional: in my opinion this should be two different unit tests as or we should use `pytest.mark.parametrize` (maybe merge with the next test then as well)

**Review by wochinge — APPROVED** [review]
Thanks for jumping on this 🚀 
Left a couple of comments but nothing blocking 👍

File: rasa/core/channels/slack.py (hunk 15 above)
Comment (tmbo) [reply rasa/core/channels/slack.py:187]: copied this from the slack docs :joy:

File: rasa/core/channels/slack.py (hunk 17 above)
Comment (tmbo) [reply rasa/core/channels/slack.py:432]: yes I'll create another PR against master :+1:

File: rasa/core/channels/slack.py (hunk 19 above)
Comment (tmbo) [reply rasa/core/channels/slack.py:464]: I don't think that is necessary, the user won't see this response as it is returned to slack in the most likely scenario when they are using the wrong signature

File: rasa/core/channels/slack.py (hunk 20 above)
Comment (tmbo) [reply rasa/core/channels/slack.py:465]: I don't think it is common to use that code for this reason as the response would need to contain an authenticate header.

Comment (wochinge) [reply rasa/core/channels/slack.py]: Mhm, but this way we can tell the client that they have to authenticate, right?

> Although the HTTP standard specifies "unauthorized", semantically this response means "unauthenticated". That is, the client must authenticate itself to get the requested response.

Comment (tmbo) [reply rasa/core/channels/slack.py]: this isn't a standard authentication though so there is no way to send a proper authenticate header back to the requester. anyways, this is a response to slack, so doesn't really matter and a 400 as a general user error should be fine

File: rasa/core/channels/slack.py (hunk 18 above)
Comment (tmbo) [reply rasa/core/channels/slack.py:439]: very good point :+1: added this as a check (and some tests) 🚀
